core/src/databases_store/gcs.rs

```diff
@@ -0,0 +1,214 @@
+        limit_offset: Option<(usize, usize)>,
+    ) -> Result<(Vec<Row>, usize)> {
+        let rows = Self::get_rows_from_csv(table).await?;
+        match limit_offset {
+            Some((limit, offset)) => {
+                let total = rows.len();
+                let rows = rows.into_iter().skip(offset).take(limit).collect();
+                Ok((rows, total))
+            }
+            None => {
+                let total = rows.len();
+                Ok((rows, total))
+            }
+        }
+    }
+
+    async fn batch_upsert_table_rows(
+        &self,
+        table: &Table,
+        rows: &Vec<Row>,
+        truncate: bool,
+    ) -> Result<()> {
+        let mut now = utils::now();
+        let mut new_rows = rows.clone();
+        let mut merge_rows_duration = 0;
+
+        // We need to merge the existing rows with the new rows based on the row_id.
+        if !truncate {
+            // We download all the rows from the CSV and merge them with the new rows.
+            // This is not super efficient if we get rows one by one but it's simple and works.
+            // Non-truncate upserts are < 4% of our total upserts.
+
+            let previous_rows = Self::get_rows_from_csv(table).await?;
+
+            // Use Hashmaps with the row_id as the key.
+            let previous_rows_map = previous_rows
+                .into_iter()
+                .map(|r| (r.row_id.clone(), r.clone()))
+                .collect::<HashMap<_, _>>();
+
+            let new_rows_map = new_rows
+                .into_iter()
+                .map(|r| (r.row_id.clone(), r.clone()))
+                .collect::<HashMap<_, _>>();
+
+            // Merge the two maps to get the final rows.
+            let merged_rows_map = previous_rows_map
+                .into_iter()
+                .chain(new_rows_map.into_iter())
+                .collect::<HashMap<_, _>>();
+
+            new_rows = merged_rows_map.into_values().collect();
+
+            merge_rows_duration = utils::now() - now;
+            now = utils::now();
+        }
+
+        let rows = Arc::new(new_rows.clone());
+
+        // Compute the schema from the rows.
+        let schema = TableSchema::from_rows_async(rows).await?;
+
+        let schema_duration = utils::now() - now;
+        now = utils::now();
+
+        // Write the rows to the CSV file.
+        Self::write_rows_to_csv(table, &schema, &new_rows).await?;
+        let write_rows_to_csv_duration = utils::now() - now;
+
+        info!(
+            truncate,
+            rows_count = new_rows.len(),
+            duration = merge_rows_duration + schema_duration + write_rows_to_csv_duration,
+            merge_rows_duration,
+            schema_duration,
+            write_rows_to_csv_duration,
+            table_id = table.unique_id(),
+            "DSSTRUCTSTAT [upsert_rows CSV] operation completed"
+        );
+
+        Ok(())
+    }
+
+    async fn delete_table_rows(&self, table: &Table) -> Result<()> {
```

Comment:
nit: the name `delete_table_rows` is a attached to the tables_rows impl, and `delete_table_data` might be a bit more correct is the general sense.